frontend/src/lib/keybind.ts

```diff
@@ -0,0 +1,246 @@
+  });
+
+  if (currentResultIndex != null) {
+    newIndex = currentResultIndex + direction;
+    if (0 > newIndex || newIndex >= results.length) {
+      newIndex = currentResultIndex;
+    }
+  }
+
+  const nextResult = results[newIndex];
+
+  // Get the anchor element and focus it
+  (nextResult.getElementsByClassName('result-main-link')[0] as HTMLElement)?.focus();
+};
+
+/**
+ * Wrapper for {@link navigateResults} that will focus on the next result
+ */
+const focusNextResult = () => navigateResults(1);
+
+/**
+ * Wrapper for {@link navigateResults} that will focus on the previous result
+ */
+const focusPrevResult = () => navigateResults(-1);
+
+/**
+ * Focus the first result (if any)
+ */
+const focusMainResult = () => {
+  const results = document.getElementsByClassName('result-main-link');
+  if (results.length > 0) {
+    (results[0] as HTMLElement).focus();
+  }
+};
+
+/**
+ * Focus the search bar
+ */
+const focusSearchBar = () => {
+  (document.getElementById('searchbar') as HTMLInputElement)?.select();
+};
+
+/**
+ * Scroll to the top of the window and reset focus
+ */
+const scrollToTop = () => {
+  window.scrollTo({ top: 0, behavior: 'smooth' });
+  if (document.activeElement instanceof HTMLElement) {
+    document.activeElement.blur(); // reset focus
+  }
+};
+
+/**
+ * Redirect to the currently focused result
+ *
+ * @param e - The triggering {@link KeyboardEvent}
+ */
+const openResult = (e: KeyboardEvent) => {
+  if ((e.target as HTMLElement).className.includes('result-main-link')) {
+    const resultLink = e.target as HTMLAnchorElement;
+    window.location.href = resultLink.href;
+  } else {
+    const results = document.getElementsByClassName('result-main-link');
+    if (results.length > 0) {
+      window.location.href = (results[0] as HTMLAnchorElement).href;
+    }
+  }
+};
+
+/**
+ * Open the currently focused result in a new tab
+ *
+ * @remarks
+ * Requires pop-ups to be allowed for the window
+ *
+ * @param e - The triggering {@link KeyboardEvent}
+ */
+const openResultInNewTab = (e: KeyboardEvent) => {
+  // Ensure target event is a result url
+  if ((e.target as HTMLElement).className.includes('result-main-link')) {
+    const resultLink = e.target as HTMLAnchorElement;
+    window.open(resultLink.href, '_blank', 'noopener');
+  }
+};
+
+/**
+ * Do a domain search using the domain of the currently focused result
+ *
+ * @param e - The triggering {@link KeyboardEvent}
+ */
+const domainSearch = (e: KeyboardEvent) => {
+  // Ensure target event is a result url
+  if ((e.target as HTMLElement).className.includes('result-main-link')) {
+    const searchBar = document.getElementById('searchbar') as HTMLInputElement;
+    if (searchBar) {
+      // Parse the result url, pull out the hostname then resubmit search
+      const resultLink = new URL((e.target as HTMLAnchorElement)?.href);
+      searchBar.value += ` site:${resultLink.hostname}`;
+      searchBar.form?.submit();
+    }
+  }
```

Comment:
should not directly reference the dom elements but instead call related `.userQuery()` and ``search(`${query} site:${resultLink.hostname}`)`` methods. `ResultLink.svelte` should also have a `.url()` method

Reply:
can you expand on `.userQuery()` & `search(...)`? I couldn't find any `userQuery` methods anywhere and was just going to pull it from `PageData` and pass it in a wrapper. As for `search` I was going to use [this](https://github.com/StractOrg/stract/blob/main/frontend/src/lib/search.ts#L100) if that's what you were referring to. Wasn't sure if these are existing need to be built out.

Reply:
Yea I think it would be easiest to create those as new methods in `Searchbar.svelte`

```ts
export const userQuery = () => lastRealQuery;
export const search = (q: string) => {
  if (formElem && inputElem) {
    inputElem.value = q;
    formElem.submit();
  }
};
```
with an added `formElem` bind
```ts
let formElem: HTMLFormElement | undefined;
...
<form
  action="/search"
  class="flex w-full justify-center"
  id="searchbar-form"
  bind:this={formElem}
  method={$postSearchStore ? 'POST' : 'GET'}
>
```

The search function in `search.ts` effectively just calls the api and returns the results. Calling the search function in `Searchbar.svelte` would ensure that the user interface gets updated correctly with the new results.

`domainSearch` can then be something like
```ts
const domainSearch = (_e: KeyboardEvent, refs: Refs) => {
  const focusedResult: Result | undefined = refs.results.find((result) => result.hasFocus());

  if (refs.searchbar && focusedResult) {
    const query: string = refs.searchbar.userQuery();
    const url: string = focusedResult.getMainLink().url();
    const domain = new URL(url).hostname;

    const domainQuery = `site:${domain}`;

    if (query.includes(domainQuery)) {
      refs.searchbar.search(query);
    } else {
      refs.searchbar.search(`${query} site:${domain}`);
    }
  }
}
```

Where `getMainLink` would be a new method in `Result.svelte`
```ts
export const getMainLink = () => mainLink
```
and `url` would be a new method in `ResultLink.svelte`
```ts
export const url = () => href;
```